ext/standard/exec.c

```diff
@@ -547,6 +556,19 @@ PHP_FUNCTION(shell_exec)
 	if (ret && ZSTR_LEN(ret) > 0) {
 		RETVAL_STR(ret);
 	}
+ }
+/* }}} */
+
+#else
+
+void php_shell_exec_ex(INTERNAL_FUNCTION_PARAMETERS);
```

Comment:
This reads really odd - if WASI is active, you only use a forward declaration. I would rather do it like:

```
#ifndef PHP_WASI
static  void php_shell_exec_ex(INTERNAL_FUNCTION_PARAMETERS) /* {{{
{
…
}
#else
static  void php_shell_exec_ex(INTERNAL_FUNCTION_PARAMETERS) /* {{{
{
    php_wasi_shell_exec_ex(INTERNAL_FUNCTION_PARAMETERS_PASSTHRU);
}
#endif // PHP_WASI
```

And then have your new files implement the `php_wasi_*` variant(s).

But considering that doesn't do anything on WASI, this function should also just not even exist on that platform.